### PR TITLE
Detect liars

### PR DESCRIPTION
**Pull request by antoinevastel**
I added new features to detect liars or people with inconsistencies in their fingerprint, particularly in their screen resolution, their languages or their OS. I followed the advices given in contributing.md.

Regards,

Antoine VASTEL

### CODE REVIEW

**Comment by Valve** [comment]
Wow, thanks! Will merge a bit later

File: package.json [hunk 1]
```diff
@@ -8,9 +8,12 @@
   },
   "devDependencies": {
     "eslint": "^0.14.1",
-    "gulp": "^3.8.11",
+    "gulp": "~3.9.0",
     "gulp-eslint": "^0.4.2",
-    "gulp-uglify": "^1.1.0"
+    "gulp-uglify": "^1.1.0",
+    "gulp-util": "~3.0.6",
+    "jshint-stylish": "~2.0.1",
```
Comment (Valve) [inline package.json]: Just curious, why you added this, since we already have eslint.

File: fingerprint2.js [hunk 2]
```diff
@@ -192,6 +193,24 @@
       }
       return keys;
     },
+    hasLiedLanguagesKey: function(keys){
+      if(!this.options.excludeHasLiedLanguages){
+        keys.push(this.getHasLiedLanguages());
+      }
+      return keys;
+    },
+    hasLiedResolutionKey: function(keys){
```
Comment (Valve) [inline fingerprint2.js]: You defined this method but never call it.

File: fingerprint2.js [hunk 3]
```diff
@@ -192,6 +193,24 @@
...
+      if(!this.options.excludeHasLiedResolution){
+        keys.push(this.getHasLiedResolution());
+      }
+      return keys;
+    },
+    hasLiedOsKey: function(keys){
```
Comment (Valve) [inline fingerprint2.js]: Same here

File: fingerprint2.js [hunk 4]
```diff
@@ -591,6 +610,97 @@
...
+    },
+    getHasLiedOs: function(){
+      var userAgent = navigator.userAgent;
+      var oscpu = navigator.oscpu;
+      var platform = navigator.platform;
+      var os;
+      //We extract the OS from the user agent (respect theorder of the if else if statement)
+      if(userAgent.toLowerCase().indexOf("windows phone") >= 0){
+        os = "Windows Phone";
+      }else if(userAgent.toLowerCase().indexOf("win") >= 0){
```
Comment (Valve) [inline fingerprint2.js]: please add spaces to be `} else`.

File: fingerprint2.js [hunk 5]
```diff
@@ -591,6 +610,97 @@
...
+      }else{
+        mobileDevice = false;
+      }
+
+      if(mobileDevice && os !== "Windows Phone" && os !== "Android" && os !== "iOS" && os !== "Other"){
+        return true;
+      }
+
+      //We compare oscpu with the os extracted from the ua
+      if(oscpu !== undefined){
```
Comment (Valve) [inline fingerprint2.js]: A better way would be 
```javascript
typeof oscpu != "undefined"
```

File: fingerprint2.js [hunk 6]
```diff
@@ -591,6 +610,97 @@
...
+        if(oscpu.toLowerCase().indexOf("win") >= 0 && os !== "Windows" && os !== "Windows Phone"){
+          return true;
+        }else if(oscpu.toLowerCase().indexOf("linux") >= 0 && os !== "Linux" && os !== "Android"){
```
Comment (Valve) [inline fingerprint2.js]: please add space between curly and else

File: README.md [hunk 7]
```diff
@@ -84,6 +84,9 @@ python -m SimpleHTTPServer
 17. WebGL fingerprinting
 18. Plugins (IE included)
 19. Is AdBlock installed or not
+20. Has the user lied on its languages
```
Comment (Valve) [inline README.md]: maybe 'tampered with' is better?

**Comment by antoinevastel** [comment]
For the dependencies in the json file it's a mistake. I had never used gulp. Npm has added things into the file. I will take into consideration your remarks :)

**Comment by antoinevastel** [comment]
I made the modifications.

**Comment by Valve** [comment]
Sorry for not pointing it out sooner, but I've just noticed that you're using a  lot of `String#indexOf`, which is not supported in <=IE8. Please add a polyfill, which uses native String.prototype.indexOf if available, and  implementation, if not. Thanks!

File: fingerprint2.js [hunk 8]
```diff
@@ -591,6 +612,97 @@
...
+
+      if(mobileDevice && os !== "Windows Phone" && os !== "Android" && os !== "iOS" && os !== "Other"){
+        return true;
+      }
+
+      //We compare oscpu with the os extracted from the ua
+      if(typeof oscpu !== "undefined"){
+        if(oscpu.toLowerCase().indexOf("win") >= 0 && os !== "Windows" && os !== "Windows Phone"){
+          return true;
+        }else if(oscpu.toLowerCase().indexOf("linux") >= 0 && os !== "Linux" && os !== "Android"){
```
Comment (Valve) [inline fingerprint2.js]: please update spacing here too. Sorry for this, I will add an eslint rule to verify this automatically.

**Comment by antoinevastel** [comment]
Done

**Comment by antoinevastel** [comment]
I added the polyfill at the begining of the file because i didn't know where i had to put it. Feel free to move it ;)

**Comment by Valve** [comment]
Thanks!

**Pull request merged**